Clarify documentation workflows for assemblies with UseCompilerGeneratedDocXmlFile (#121271)
dotnet/runtime: docs/coding-guidelines/adding-api-guidelines.md

**For assemblies with `<UseCompilerGeneratedDocXmlFile>false</UseCompilerGeneratedDocXmlFile>`:**
- The [dotnet-api-docs](https://github.com/dotnet/dotnet-api-docs) repo is the source of truth for documentation.
- Triple-slash comments in source code are synced to dotnet-api-docs **only once** for newly introduced APIs.
- After the initial sync, all subsequent documentation updates must be made directly in the dotnet-api-docs repo.
- Older assemblies, including System.Private.CoreLib, typically follow this workflow.

**For assemblies without this property (or with it set to `true`, which is the default):**
- Source comments in this repo are the source of truth for documentation.
- After your change is merged, we will eventually port them to the dotnet-api-docs repo, where we will review them for language and proper style (For more information, see the [API writing guidelines](https://github.com/dotnet/dotnet-api-docs/wiki)).
- Once the dotnet-api-docs change is merged, your comments will start showing up in the official API documentation at https://learn.microsoft.com, and later they'll appear in IntelliSense in Visual Studio and Visual Studio Code.

> - Once the documentation is official, any subsequent updates to it must be made directly in https://github.com/dotnet/dotnet-api-docs/. It's fine to make updates to the triple slash comments later, they just won't automatically flow into the official docs.

Reviewer: Line 62 seems to contradict Line 60. I understand 62 to say updates will not be merged to dotnet-api-docs and 60 says they will be merged. I'm guessing line 60 is the accurate one or there is still some subtlety I am missing.

New public APIs must be documented with triple-slash comments on top of them. Visual Studio automatically generates the structure for you when you type `///`.

[API writing guidelines](https://github.com/dotnet/dotnet-api-docs/wiki) has information about language and proper style for writing API documentation.
If your new API or the APIs it calls throw any exceptions, those need to be manually documented by adding the `<exception></exception>` elements.

> After your change is merged, we will eventually port them to the [dotnet-api-docs](https://github.com/dotnet/dotnet-api-docs) repo. Once the dotnet-api-docs change

Reviewer 2: ```suggestion
After your change is merged, we will eventually port them to the [dotnet-api-docs](https://github.com/dotnet/dotnet-api-docs) repo. The tools used for this port live in [api-docs-sync](https://github.com/dotnet/api-docs-sync) repo. Once the dotnet-api-docs change
```

is merged, your comments will start showing up in the official API documentation at https://learn.microsoft.com, and later they'll appear in IntelliSense
in Visual Studio and Visual Studio Code.

The rest of the documentation workflow depends on whether the assembly has the `UseCompilerGeneratedDocXmlFile` property set in its project file:

**For libraries without this property (or with it set to `true`, which is the default):**
- Source comments in this repo are the source of truth for documentation.
- Triple-slash comments in source code are synced to dotnet-api-docs periodically.
- More recently introduced libraries typically follow this workflow.

**For libraries with `<UseCompilerGeneratedDocXmlFile>false</UseCompilerGeneratedDocXmlFile>`:**
- The [dotnet-api-docs](https://github.com/dotnet/dotnet-api-docs) repo is the source of truth for documentation.
- Triple-slash comments in source code are synced to dotnet-api-docs **only once** for newly introduced APIs. After the initial sync, all subsequent documentation
updates must be made directly in the dotnet-api-docs repo.

> - It's fine to make updates to the triple-slash comments later, they just won't automatically flow into the official docs. Copilot can help with porting the changes

Reviewer 3: For small changes where it can apply a diff, sure.  If you're adding a lot of docs, or applying docs to something that was missed in many APIs (return values, generic parameters, etc) it doesn't do well, especially when it needs to decode doc ID from C# to understand overloads.  If it does make a mistake you end up having to fix it anyway.  If you wanted to link a document you could link to https://github.com/dotnet/api-docs-sync/blob/main/docs/PortToDocs.md which contains the small set of steps to do this porting.

For copilot, it's important that it understands the style guidelines in https://github.com/dotnet/dotnet-api-docs/wiki and https://github.com/github/awesome-copilot/blob/main/prompts/csharp-docs.prompt.md

> in triple-slash comments to dotnet-api-docs. [dotnet-api-docs](https://github.com/dotnet/api-docs-sync/blob/main/docs/PortToDocs.md) tools work better to ports of large changes.

Reviewer 2: ```suggestion
in triple-slash comments to dotnet-api-docs. [PortToDocs](https://github.com/dotnet/api-docs-sync/blob/main/docs/PortToDocs.md) tool works better for ports of large changes.
```

> - Older libraries typically follow this workflow. Libraries in this mode can work towards a better workflow in the future by using api-docs-sync tool to port back docs to source, then removing the `UseCompilerGeneratedDocXmlFile` property.

Reviewer 2: ```suggestion
- Older libraries typically follow this workflow. Libraries in this mode can work towards a better workflow in the future by using api-docs-sync tools to port back docs to source, then removing the `UseCompilerGeneratedDocXmlFile` property.
```
